### PR TITLE
safely infer device type + docstrings + nits

### PR DESCRIPTION
**Pull request by nmacchioni**
we can't safely infer device type from the generic method `Benchmarker.benchmark(fn, fn_args, fn_kwargs)` if there is no device type in `fn_args` or `fn_kwargs`, or if there are multiple device types. in this case, we'll raise `ValueError` and redirect the caller to a device-specific implementation (see https://github.com/pytorch/pytorch/pull/133290 for an example of this)

+ fixed up some docstrings
+ small nits

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #133378
* #133287
* #133154
* #133153
* #133121
* #133058
* #132966
* #132959
* #132958
* __->__ #133286

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

**Comment by nmacchioni** [comment]
@eellison what do you think about erroring out when we can't safely infer device type?

File: torch/_inductor/runtime/benchmarking.py [hunk 1]
```diff
@@ -49,50 +48,76 @@ def benchmark(
...
+        - **kwargs: The benchmarking implementation's kwargs.
 
         Returns:
         - The runtime of `fn(*fn_args, **fn_kwargs)`, in milliseconds.
         """
-        if is_cpu_device(list(fn_args) + list(fn_kwargs.values())):
-            return self.benchmark_cpu(lambda: fn(*fn_args, **fn_kwargs), **kwargs)
-        return self.benchmark_gpu(lambda: fn(*fn_args, **fn_kwargs), **kwargs)
+        inferred_device = None
+        for arg_or_kwarg in fn_args + tuple(fn_kwargs.items()):
```
Comment (eellison) [inline torch/_inductor/runtime/benchmarking.py:71]: `tuple(fn_kwargs.items()` will iterate over a tuples of (key, value) so we'll never see Tensros here. also, better not to materialize a tuple, you can do `itertools.chain(fn_args, fn_kwargs.keys(), fn_kwargs.values())`

**Review by eellison — COMMENTED** [review]
hmm, we could error out, I guess I would probably just keep the current behavior of benchmark gpu if we cant figure it out.

will defer to you though

**Comment by nmacchioni** [comment]
Will error out for now, I think this can reveal some true problems (i.e. the problem we already saw in `test_inductor_utils.py`.

File: torch/_inductor/runtime/benchmarking.py (hunk 1 above)
Comment (nmacchioni) [reply torch/_inductor/runtime/benchmarking.py:71]: yep, should be `fn_kwargs.values()`. Will use chain

**Comment by nmacchioni** [comment]
Cc @eellison for review, going to land part of stack

**Review by eellison — APPROVED** [review]

**Pull request merged**